Feat esp encrypted img/adds rsa ds peripheral support

Ashish285 opened the pull request:
# Checklist

- [ ] Component contains License
- [ ] Component contains README.md
- [ ] Component contains idf_component.yml file with `url` field defined
- [ ] Component was added to [upload job](https://github.com/espressif/idf-extra-components/blob/master/.github/workflows/upload_component.yml#L18)
- [ ] Component was added to [build job](https://github.com/espressif/idf-extra-components/blob/master/test_app/CMakeLists.txt#L8)
- [ ] _Optional:_ Component contains unit tests
- [ ] CI passing

# Change description
_Please describe your change here_

mahavirj commented on esp_encrypted_img/include/esp_encrypted_img.h:
```diff
@@ -68,6 +68,10 @@ typedef void *esp_decrypt_handle_t;
 #if defined(CONFIG_PRE_ENCRYPTED_OTA_USE_RSA)
+#if defined(CONFIG_PRE_ENCRYPTED_RSA_USE_DS)
+    void *ds_data;                         /*!< Data context for DS */
+    int rsa_padding;                       /*!< Padding type for RSA decryption */
```
I think we can remove this option

mahavirj commented on esp_encrypted_img/include/esp_encrypted_img.h:
```diff
@@ -68,6 +68,10 @@ typedef void *esp_decrypt_handle_t;
 typedef struct {
 #if defined(CONFIG_PRE_ENCRYPTED_OTA_USE_RSA)
+#if defined(CONFIG_PRE_ENCRYPTED_RSA_USE_DS)
+    void *ds_data;                         /*!< Data context for DS */
```
This could be purely dynamic as well - if `ds_data` is set then use DS peripheral approach else tradition RSA approach. Then we can remove `CONFIG_PRE_ENCRYPTED_RSA_USE_DS`

Ashish285 replied:
I am not sure about making it dynamic and creating this fallback. If the choice is made during runtime dynamically, then it might be possible (intentionally or not) to fall back to a less secure option. If the choice is compile time, then we can guarantee the security, and if it fails the OTA will not succeed. Let me know what are your thoughts.

mahavirj commented on esp_encrypted_img/idf_component.yml:
```diff
@@ -4,3 +4,5 @@ url: https://github.com/espressif/idf-extra-components/tree/master/esp_encrypted
   idf:
     version: ">=4.4"
+  espressif/esp_secure_cert_mgr:
+    version: '*'
```
Lets set rule to allow only minor or patch version. Major version may break the API.

Ashish285 replied:
Added in version check

mahavirj commented:
Let's bump up the minor component version and add IDF version dependency for newly added feature

Ashish285 commented:
> Let's bump up the minor component version and add IDF version dependency for newly added feature

Sure, updated the component version and added in latest IDF version dependency. Thanks

mahavirj commented on esp_encrypted_img/idf_component.yml:
```diff
@@ -1,6 +1,8 @@
 dependencies:
   idf:
-    version: ">=4.4"
+    version: ">=5.5"
```
This will lock in the component for IDF >= 5.5 release. We should support prior releases and take an exception for newly added feature only

Ashish285 commented on esp_encrypted_img/project_include.cmake:
```diff
@@ -23,3 +25,24 @@ function(create_esp_enc_img input_file key_file output_file app)
     add_dependencies(enc_bin_target_${name} gen_project_binary)
 endfunction()
+
+# function(create_esp_enc_img_secure_cert_data target_chip input_key input_key_algo)
```
This is currently commented out because esp_secure_cert_mgr currently doesn't allow creating secure cert partition without the port. Once we add in the functionality we should be able to create the secure partition internally and the user doesn't have to create one by themselves.

Harshal5 replied:
Oh okay got it. Thanks!

Ashish285 commented on esp_encrypted_img/src/esp_encrypted_img.c:
```diff
@@ -417,25 +484,20 @@ static int derive_gcm_key(const char *data, esp_encrypted_img_t *handle)
-        mbedtls_mpi_free(&device_private_mpi);
         goto exit;
     }
-    memset(&device_private_mpi, 0, sizeof(mbedtls_mpi));
```
Not really required as `mbedtls_mpi_free` internally does `mbedtls_mpi_zeroize_and_free`

Ashish285 commented on esp_encrypted_img/src/esp_encrypted_img.c:
```diff
@@ -95,6 +109,53 @@ typedef struct esp_encrypted_img_handle esp_encrypted_img_t;
+                             (unsigned char *)handle->gcm_key, &olen, GCM_KEY_SIZE,
+                             mbedtls_esp_random, NULL);
+    if (ret != 0) {
+        ESP_LOGE(TAG, "failed\n  ! mbedtls_pk_decrypt returned -0x%04x\n", (unsigned int) - ret);
```
Yeah, I think the idea with logging here is that in case we have mbedtls error code, then it prints out `failed` along with the mbedtls error code, if not, then it just prints out a generic failure statement. Maybe we can revisit it in future and merge these into one pattern
